### PR TITLE
Fix bogus assertion tripped by force-executed tasks

### LINKED ISSUE #104580: Tasks submitted to shutting down fixed executors are sometimes neither executed nor rejected
AFAICT it seems that fixed executors (with bounded queues) sometimes just drop tasks on the floor after shutdown, rather than rejecting them as one might expect. For a demonstration test case, see `testFixedBoundedRejectOnShutdown` added (albeit muted) in https://github.com/elastic/elasticsearch/pull/104581 - this test fails for me whenever it submits a force-executed task.

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today if you submit a force-executing task to a fixed executor with an
unbounded queue after the executor has been shut down then the
`EsAbortPolicy` will incorrectly throw an `IllegalStateException` which
ends up tripping an assertion. This is a legitimate thing to happen, so
this commit introduces a different rejection handler to deal with it. It
also introduces a check for a shut-down bounded-queue executor on
rejection to make sure we don't force the task onto a queue that's never
going to be processed.

Closes #104580

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java [hunk 1]
```diff
@@ -8,26 +8,27 @@
...
+                        throw new IllegalStateException("forced execution, but got interrupted", e);
+                    }
                 }
-                try {
-                    ((SizeBlockingQueue<Runnable>) queue).forcePut(r);
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    throw new IllegalStateException("forced execution, but got interrupted", e);
+
+                if (executor.isShutdown() == false) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java:29]: Another possible fix here is to use a different rejection handler if the executor has an unbounded queue.

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java:29]: If the executor is shutting down, then this condition evaluates to false, and we fall through to reject execution below, right? Yet we may have already forced the task onto the queue, so don't we need to try to take it off before claiming it is rejected, like in EsExecutors.ForceQueuePolicy? Actually why do we even have two rejected execution handlers at all?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java:29]: If we forced the task onto the queue then we don't get here, we already called `return;`. But yes, it looks like there's a bug here too, see https://github.com/elastic/elasticsearch/issues/104580#issuecomment-1900462181.

> Actually why do we even have two rejected execution handlers at all?

Why indeed.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java]: AFAICT the different handlers correspond with the different types of queue in use. Therefore it seems we should have another handler for the unbounded `LinkedTransferQueue` that causes this failure - see 212e813dc8d7df7557b6f30a1bf66409440a72f5.

**Review by rjernst — APPROVED** [review]
This looks good.

I wonder if as a followup we could move EsAbortPolicy into EsExecutors alongside these other two rejected execution handlers. The only uses of EsAbortPolicy I see outside of EsExecutors is PrioritizedEsThreadPoolExecutor (by calling EsThreadPoolExecutor ctor which does not take a rejected execution handler), and Scheduler which creates a SafeScheduledThreadPoolExecutor. Both of these uses have queues that are not SizeBlockingQueue, so the handler would seem to be pointless (or incorrect).

**Pull request merged**